test/common/run-tests

```diff
@@ -292,147 +306,134 @@ def detect_tests(test_files, image, opts):
     changed_tests = get_affected_tests(opts.test_dir, opts.base, test_files)
-    (serial_tests, parallel_tests) = detect_tests(test_files, image, opts)
+    (serial_tests, parallel_tests) = detect_tests(test_files, opts)
+    serial_tests_len = len(serial_tests)
+    parallel_tests_len = len(parallel_tests)
 
-    print("1..{0}".format(len(parallel_tests) + len(serial_tests)))
+    print(f"1..{serial_tests_len + parallel_tests_len}")
     flush_stdout()
 
-    if serial_tests and opts.list:
-        # Just build one batch for listing
-        batch_tests[0] = {"working": None, "tests": [], "time": 0}
-        for test in serial_tests:
-            test.assign_machine(0, "listdummy", "listdummy")
-            batch_tests[0]["tests"].append(test)
-
-    if serial_tests and not opts.list:
-        if opts.machine:
-            batch_tests[0] = {"working": None, "tests": [], "time": 0}
-            for test in serial_tests:
-                batch_tests[0]["tests"].append(test)
-                test.assign_machine(0, opts.machine, opts.browser)
-        else:
-            batch_size = len(serial_tests) // opts.batches
-
-            for i in range(opts.batches):
-                m = GlobalMachine(restrict=not opts.enable_network, cpus=opts.nondestructive_cpus, memory_mb=opts.nondestructive_memory_mb)
-                batch_tests[i] = {"working": None, "tests": [], "time": 0, "machine": m}
-                ssh_address = f"{m.machine.ssh_address}:{m.machine.ssh_port}"
-                web_address = f"{m.machine.web_address}:{m.machine.web_port}"
-
-                if i == opts.batches - 1:  # Last machine needs to resolve the rest
-                    batch = serial_tests[batch_size * i:]
-                else:
-                    batch = serial_tests[batch_size * i: batch_size * i + batch_size]
-
-                for test in batch:
-                    batch_tests[i]["tests"].append(test)
-                    test.assign_machine(i, ssh_address, web_address)
-
     running_tests = []
-    serial_tests_len = len(serial_tests)
-    serial_remaining = sum([len(batch_tests[x]["tests"]) for x in batch_tests])
-    while serial_remaining or parallel_tests or running_tests:
+    global_machines = []
+
+    # create appropriate number of serial machines; assume a coarse approximation that a typical destructive (parallel)
+    # test takes twice the time of a typical serial test; the goal is to not over-parallelize serial tests if there
+    # are lots of parallel ones (to detect inter-test dependencies), but use maximum parallelization if the test suite
+    # only has a few or even no parallel tests
+    if serial_tests_len > 0:
+        num_global = max(1, round(serial_tests_len / (serial_tests_len + 2 * parallel_tests_len) * opts.jobs))
+    else:
+        num_global = 0
+    logging.debug(f"allocating {num_global} global machines for {serial_tests_len} serial and {parallel_tests_len} parallel tests")
```

Comment:
I kinda disagree with the logic here.  I'm not sure if semi-reliable detection of inter-test dependencies is a desirable characteristic or not, but I feel like we could accomplish that by randomising the queue.

How about this (rather simple) approach:
 - ignore parallel tests at first
 - start `n` serial job runners for `n = min(len(serial_tests), opts.jobs)`, and have them consume `@nd` tests from the queue
 - as each serial runner finishes its last job, have it switch over to running parallel tests

This feels like a better way to keep all cores as busy as possible, without making any generalized guess about the relative required running time of the parallel vs. serial tests.

Reply:
One more point: the names "serial" and "parallel" are awful here.  I wonder if we could get rid of them?

Reply:
FTR: I considered this, and discarded it as the original rationale still seems to make sense. But indeed, let's do this simple approach, check our statistics in a few weeks, and let the data talk. This is super-easy to tweak now.